core/commonMain/src/implementations/immutableMap/TrieNode.kt

```diff
@@ -436,6 +439,81 @@ internal class TrieNode<K, V>(
         return this
     }
 
+    private fun collisionPutAll(otherNode: TrieNode<K, V>, intersectionCounter: DeltaCounter): TrieNode<K, V> {
+        assert(nodeMap == 0)
+        assert(dataMap == 0)
+        assert(otherNode.nodeMap == 0)
+        assert(otherNode.dataMap == 0)
+        var res = this
+        for(i in 0 until otherNode.buffer.size step ENTRY_SIZE) {
+            val k = otherNode.keyAtIndex(i)
+            val v = otherNode.valueAtKeyIndex(i)
+
+            val mres = res.collisionPut(k, v)
+            if (mres == null || mres.sizeDelta == 0) intersectionCounter += 1
+            res = mres?.node ?: res
+        }
+        return res
+    }
+
+    private fun putAllFromOtherNodeCell(other: TrieNode<K, V>,
+                                        positionMask: Int,
+                                        shift: Int,
+                                        intersectionCounter: DeltaCounter): TrieNode<K, V> {
+        return when {
+            other.hasNodeAt(positionMask) -> {
+                putAll(
+                        other.nodeAtIndex(other.nodeIndex(positionMask)),
+                        shift + LOG_MAX_BRANCHING_FACTOR,
+                        intersectionCounter
+                )
+            }
+            other.hasEntryAt(positionMask) -> {
+                val keyIndex = other.entryKeyIndex(positionMask)
+                val key = other.keyAtIndex(keyIndex)
+                val value = other.valueAtKeyIndex(keyIndex)
+                val newNodeResult = put(
+                        key.hashCode(),
+                        key,
+                        value,
+                        shift + LOG_MAX_BRANCHING_FACTOR
+                )
+                if (newNodeResult == null) {
+                    intersectionCounter += 1
+                    this
+                } else {
+                    if (newNodeResult.sizeDelta == 0) {
+                        /* if put didn't change the size, there was a replacement */
+                        intersectionCounter += 1
+                    }
+                    newNodeResult.node
+                }
+            }
+            else -> this
+        }
+    }
+
+    private fun calculateSize(): Int {
+        if (nodeMap == 0) return buffer.size / ENTRY_SIZE
+        val numValues = dataMap.countOneBits()
+        var result = numValues
+        for(i in (numValues * ENTRY_SIZE) until buffer.size) {
+            result += nodeAtIndex(i).calculateSize()
+        }
+        return result
+    }
+
+    private fun elementsEquals(otherNode: TrieNode<K, V>): Boolean {
+        if (this === otherNode) return true
```

Comment:
`newNode` is a newly created object and cannot be `===` to `this` node or `otherNode`.

Reply:
This comment was relevant in immutable implementation